File: src/commands/projects.ts
import { Command } from "@effect/cli"
import { Console, Effect } from "effect"
import { ConfigService } from "../services/config.js"
import { dockerContainerOf } from "../schema/config.js"
import { bold, dim, blue } from "../fmt.js"

// ---------------------------------------------------------------------------
// ship projects — list registered projects
// ---------------------------------------------------------------------------

export const projectsCommand = Command.make(
  "projects",
  {},
  () =>
    Effect.gen(function* () {
      const config = yield* ConfigService
      const shipConfig = yield* config.loadConfig()
      const entries = Object.entries(shipConfig.projects)

      if (entries.length === 0) {
        yield* Console.log("")
        yield* Console.log(`  ${dim("No projects registered.")}`)
        yield* Console.log(`  ${dim("Register one with: ship init")}`)
        yield* Console.log("")
        return
      }

      const aliasWidth = Math.max(7, ...entries.map(([a]) => a.length))

      yield* Console.log("")
      yield* Console.log(
        `  ${"ALIAS".padEnd(aliasWidth)}  ${"PATH".padEnd(40)}  ${"DB CONTAINER"}`
      )
      yield* Console.log(
        `  ${dim("─".repeat(aliasWidth))}  ${dim("─".repeat(40))}  ${dim("─".repeat(16))}`
      )

      for (const [alias, project] of entries) {
        yield* Console.log(
          `  ${bold(alias.padEnd(aliasWidth))}  ${blue(project.path.padEnd(40))}  ${dockerContainerOf(project.database)}`
        )
      }

      yield* Console.log("")
      yield* Console.log(`  ${dim(`${entries.length} project${entries.length === 1 ? "" : "s"}`)}`)
      yield* Console.log("")
    }).pipe(
      Effect.catchAll((e) =>
        Console.error(`Error: ${e.message}`)
      )
    )
)

File: src/commands/projects.rs
use crate::errors::Result;
use crate::fmt::{blue, bold, dim};
use crate::services::config;
use crate::util::plural;

// ---------------------------------------------------------------------------
// ship projects — list registered projects
// ---------------------------------------------------------------------------

pub fn run() {
    if let Err(e) = run_inner() {
        eprintln!("Error: {e}");
    }
}

fn run_inner() -> Result<()> {
    let ship_config = config::load_config()?;
    let entries: Vec<_> = ship_config.projects.iter().collect();

    if entries.is_empty() {
        println!();
        println!("  {}", dim("No projects registered."));
        println!("  {}", dim("Register one with: ship init"));
        println!();
        return Ok(());
    }

    let alias_width = entries
        .iter()
        .map(|(a, _)| a.len())
        .max()
        .unwrap_or(0)
        .max(7);

    println!();
    println!(
        "  {:<aw$}  {:<40}  DB CONTAINER",
        "ALIAS",
        "PATH",
        aw = alias_width
    );
    println!(
        "  {}  {}  {}",
        dim("─".repeat(alias_width)),
        dim("─".repeat(40)),
        dim("─".repeat(16))
    );

    for (alias, project) in &entries {
        println!(
            "  {}  {}  {}",
            bold(format!("{:<aw$}", alias, aw = alias_width)),
            blue(format!("{:<40}", project.path)),
            project.database.docker_container()
        );
    }

    println!();
    println!(
        "  {}",
        dim(format!(
            "{} project{}",
            entries.len(),
            plural(entries.len())
        ))
    );
    println!();
    Ok(())
}
